main/src/ca/uwaterloo/flix/language/phase/Monomorpher.scala

```diff
@@ -688,49 +688,45 @@ object Monomorpher {
   }
 
   /**
-    * Returns the def symbol corresponding to the specialized symbol `sym` w.r.t. to the type `tpe`.
-    *
-    * The given type must be a normalized type.
+    * Returns a [[Symbol.DefnSym]] corresponding to the specialization of `sym`
+    * w.r.t. the normalized function type `tpe`.
     */
   private def specializeSigSym(sym: Symbol.SigSym, tpe: Type)(implicit ctx: Context, instances: Map[(Symbol.TraitSym, TypeConstructor), Instance], root: LoweredAst.Root, flix: Flix): Symbol.DefnSym = {
+    val defn = resolveSigSym(sym, tpe)
+    specializeDef(defn, tpe)
+  }
+
+  /**
+    * Returns a [[LoweredAst.Def]] corresponding to the resolved trait `sym`
+    * w.r.t. to the normalized function type `tpe`.
+    */
+  private def resolveSigSym(sym: Symbol.SigSym, tpe: Type)(implicit instances: Map[(Symbol.TraitSym, TypeConstructor), Instance], root: LoweredAst.Root, flix: Flix): LoweredAst.Def = {
     val sig = root.sigs(sym)
     val trt = root.traits(sym.trt)
 
-    // Find out what we're specializing the sig to by unifying with the type.
+    // Find out what instance to use by unifying with the sig type.
     val subst = ConstraintSolver2.fullyUnify(sig.spec.declaredScheme.base, tpe, Scope.Top, RigidityEnv.empty)(root.eqEnv, flix).get
-    val specialization = subst.m(trt.tparam.sym)
-    val tycon = specialization.typeConstructor.get
+    val traitType = subst.m(trt.tparam.sym)
+    val tyCon = traitType.typeConstructor.get
 
-    // Lookup the instance corresponding to this type.
-    val instance = instances((sym.trt, tycon))
-
-    val defns = instance.defs.filter {
-      d => d.sym.text == sig.sym.name
-    }
+    val instance = instances((sym.trt, tyCon))
+    val defns = instance.defs.filter(_.sym.text == sig.sym.name)
 
     (sig.exp, defns) match {
       // An instance implementation exists. Use it.
-      case (_, defn :: Nil) => specializeDef(defn, tpe)
+      case (_, defn :: Nil) => defn
```

Comment:
Why don't we specialize the def?

Reply:
now this function just resolves the def, the specializeSigSym calls this function and then calls specializeDef

Reply:
Ok. Please add something to the doc to say that the returned def is not specialized yet.

Reply:
My thought is that everything in this file that specializes something has to be called `specializeX` - that is already mostly how it is, with a few exceptions that i plan to fix. do you think that's enough or should it be written out?

Reply:
Yeah ok